java/ql/test/library-tests/dataflow/collections/B.java

```diff
@@ -1,32 +1,67 @@
 import java.util.*;
 import java.util.concurrent.*;
 import java.util.regex.*;
+import java.util.stream.*;
 
 public class B {
   static Object source() { return null; }
 
   static void sink(Object obj) { }
 
-  static Object[] storeArrayElement(Object obj) { return null; }
-  static Object storeElement(Object obj) { return null; }
-  static Object storeMapKey(Object obj) { return null; }
-  static Object storeMapValue(Object obj) { return null; }
+  static Object[] storeArrayElement(Object obj) { return new Object[] {obj}; }
 
-  static Object readArrayElement(Object obj) { return null; }
-  static Object readElement(Object obj) { return null; }
-  static Object readMapKey(Object obj) { return null; }
-  static Object readMapValue(Object obj) { return null; }
+  static Object readArrayElement(Object[] obj) { return obj[0]; }
+  static boolean readArrayElement(boolean[] obj) { return obj[0]; }
+  static byte readArrayElement(byte[] obj) { return obj[0]; }
+  static char readArrayElement(char[] obj) { return obj[0]; }
+  static short readArrayElement(short[] obj) { return obj[0]; }
+  static int readArrayElement(int[] obj) { return obj[0]; }
+  static long readArrayElement(long[] obj) { return obj[0]; }
+  static float readArrayElement(float[] obj) { return obj[0]; }
+  static double readArrayElement(double[] obj) { return obj[0]; }
+
+  static <T> List<T> storeElementList(T obj) { return List.of(obj); }
+  static <T> Enumeration<T> storeElementEnumeration(T obj) { Vector<T> v = new Vector<>(); v.add(obj); return v.elements(); }
+  static <T> NavigableSet<T> storeElementNavSet(T obj) { TreeSet<T> s = new TreeSet<>(); s.add(obj); return s; }
+  static <T> Stack<T> storeElementStack(T obj) { Stack<T> s = new Stack<>(); s.push(obj); return s; }
+  static <T> BlockingDeque<T> storeElementBlockingDeque(T obj) { LinkedBlockingDeque<T> q = new LinkedBlockingDeque<>(); q.add(obj); return q; }
+  static <T> ListIterator<T> storeElementListIterator(T obj) { return List.of(obj).listIterator(); }
+
+  static <T> T readElement(Iterable<T> obj) { return obj.iterator().next(); }
+  static <T> T readElement(Iterator<T> obj) { return obj.next(); }
+  static <T> T readElement(Spliterator<T> obj) { return null; }
+  static <T> T readElement(Stream<T> obj) { return null; }
+  static <T> T readElement(Enumeration<T> obj) { return obj.nextElement(); }
+
+  static <K,V> Map.Entry<K,V> storeMapKeyEntry(K obj) { return Map.entry(obj,null); }
+  static <K,V> Map<K,V> storeMapKey(K obj) { Map<K,V> m = new TreeMap<K,V>(); m.put(obj,null); return m; }
+
+  static <K,V> Map.Entry<K,V> storeMapValueEntry(V obj) { return Map.entry(null,obj); }
+  static <K,V> Map<K,V> storeMapValue(V obj) { Map<K,V> m = 1==1?new TreeMap<K,V>():new ConcurrentHashMap<K,V>(); m.put(null,obj); return m; }
```

Comment:
Seems quite likely javac might delete one of the arms here? Perhaps use `System.currentTimeMillis` as a source of nondeterminism?

Reply:
No, I don't believe javac does that kind of optimisation. At the very least this does make it into the db, because if you simplify it then type pruning will remove some results (since the result of `storeMapValue` is cast to a `ConcurrentMap`, which is incompatible with `TreeMap`).